airbyte-lib/airbyte_lib/caches/base.py

```diff
@@ -35,15 +46,27 @@
     from sqlalchemy.engine.reflection import Inspector
     from sqlalchemy.sql.base import Executable
 
-    from airbyte_protocol.models import ConfiguredAirbyteStream
-
     from airbyte_lib.datasets._base import DatasetBase
     from airbyte_lib.telemetry import CacheTelemetryInfo
 
 
 DEBUG_MODE = False  # Set to True to enable additional debug logging.
 
 
+STREAMS_TABLE_NAME = "_airbytelib_streams"
+
+Base = declarative_base()
+
+
+class CachedStream(Base):  # type: ignore[valid-type,misc]
+    __tablename__ = STREAMS_TABLE_NAME
+
+    stream_name = Column(String, primary_key=True)
+    source_name = Column(String)
+    table_name = Column(String)
+    catalog_metadata = Column(String)
```

Comment:
Wdyt of using a composite three-column primary key here?

```suggestion
    stream_name = Column(String, primary_key=True)
    source_name = Column(String, primary_key=True)
    table_name = Column(String, primary_key=True)
    catalog_metadata = Column(String)
```

The implications would be:

1. Dupes are allowed for different sources using the same stream name.
1. Dupes are allowed for the same stream name and some source name - if table suffix and/or suffix are different.
1. If a user configures the cache with different prefix/suffix settings, they'll basically be writing to a different namespace within the cache.
   1. Meaning, I could initialize my cache with "_test1" suffix, "_test2" suffix, etc., AirbyteLib would treat these as different namespaces with different tablesets.
1. When we get to supporting multiple sources in the same cache, we might want a dynamic default table prefix like `'{source_name}_'`. But we can tackle that later.

Reply:
The problem I'm seeing with that is that the cache api is stream name only based.

For example, you have the following entries in the _airbytelib_streams table:
| stream    | source  | table                      |
|-----------|---------|----------------------------|
| my_stream | source1 | source1_my_stream          |
| my_stream | source1 | source1_prefixed_my_stream |


If you do `cache["my_stream"]`, then what table are you getting? For this prefix case, we could decide based on the table name, by comparing the table name the cache would choose for a stream name with the table name stored in the _airbytelib_streams table. I'm going to implement this.

But what about this case:

| stream    | source  | table                      |
|-----------|---------|----------------------------|
| my_stream | source1 | source1_my_stream          |
| my_stream | source2 | source2_my_stream          |

 `cache["my_stream"]` could be either of those, and there is no way to decide. Changing this to  `cache["source1_my_stream"]` raises lots of new questions - it just doesn't feel like the right approach to me, this edge case shouldn't make it harder for the "regular" case.

Reply:
So, in summary, we should actually use the table name as primary key, because that's what needs to be unique for the current feature set.

For multi-source, I think it's OK to merge the streams and overwrite / raise in the collision case - after all, this is also what would happen in case of regular destinations.

Reply:
Nice! Yeah, I think this sounds like a great path forward. 👍